File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/memory/config/QdrantCollectionInitializer.java
package com.study.webflux.rag.infrastructure.memory.config;

import java.time.Duration;
import java.util.Map;

import lombok.extern.slf4j.Slf4j;

import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.reactive.function.client.WebClient;

import com.study.webflux.rag.infrastructure.dialogue.config.properties.RagDialogueProperties;
import reactor.core.publisher.Mono;

/**
 * Qdrant 컬렉션 존재 여부를 확인하고 필요 시 자동 생성합니다.
 */
@Slf4j
@Component
public class QdrantCollectionInitializer implements ApplicationRunner {

	private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(10);

	private final RagDialogueProperties properties;
	private final WebClient webClient;

	public QdrantCollectionInitializer(RagDialogueProperties properties,
		WebClient.Builder webClientBuilder) {
		this.properties = properties;
		var qdrant = properties.getQdrant();
		WebClient.Builder builder = webClientBuilder.baseUrl(trimTrailingSlash(qdrant.getUrl()));
		if (StringUtils.hasText(qdrant.getApiKey())) {
			builder.defaultHeader("api-key", qdrant.getApiKey());
		}
		this.webClient = builder.build();
	}

	@Override
	public void run(ApplicationArguments args) {
		var qdrant = properties.getQdrant();
		if (!qdrant.isAutoCreateCollection()) {
			log.info("Qdrant 컬렉션 자동 생성이 비활성화되어 초기화를 건너뜁니다. collection={}",
				qdrant.getCollectionName());
			return;
		}

		boolean exists = collectionExists(qdrant.getCollectionName());
		if (exists) {
			log.info("Qdrant 컬렉션이 이미 존재합니다. collection={}", qdrant.getCollectionName());
			return;
		}

		createCollection(qdrant.getCollectionName(), qdrant.getVectorDimension());
	}

	private boolean collectionExists(String collectionName) {
		return webClient.get()
			.uri("/collections/{collectionName}", collectionName)
			.exchangeToMono(response -> {
				HttpStatusCode status = response.statusCode();
				if (status.is2xxSuccessful()) {
					return Mono.just(true);
				}
				if (status.value() == 404) {
					return Mono.just(false);
				}
				return response.bodyToMono(String.class)
					.defaultIfEmpty("")
					.flatMap(body -> Mono.error(new IllegalStateException(
						"Qdrant 컬렉션 조회 실패 status=" + status.value() + " body=" + body)));
			})
			.blockOptional(REQUEST_TIMEOUT)
			.orElse(false);
	}

	private void createCollection(String collectionName, int vectorDimension) {
		Map<String, Object> payload = Map.of("vectors",
			Map.of("size", vectorDimension, "distance", "Cosine"));

		webClient.put()
			.uri("/collections/{collectionName}", collectionName)
			.contentType(MediaType.APPLICATION_JSON)
			.bodyValue(payload)
			.exchangeToMono(response -> {
				HttpStatusCode status = response.statusCode();
				if (status.is2xxSuccessful() || status.value() == 409) {
					return Mono.empty();
				}
				return response.bodyToMono(String.class)
					.defaultIfEmpty("")
					.flatMap(body -> Mono.error(new IllegalStateException(
						"Qdrant 컬렉션 생성 실패 status=" + status.value() + " body=" + body)));
			})
			.block(REQUEST_TIMEOUT);

		log.info("Qdrant 컬렉션을 준비했습니다. collection={}, vectorDimension={}",
			collectionName,
			vectorDimension);
	}

	private String trimTrailingSlash(String url) {
		if (!StringUtils.hasText(url)) {
			return "";
		}
		if (url.endsWith("/")) {
			return url.substring(0, url.length() - 1);
		}
		return url;
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/memory/config/QdrantCollectionInitializer.kt
package com.miyou.app.infrastructure.memory.config

import com.miyou.app.infrastructure.dialogue.config.properties.RagDialogueProperties
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.http.HttpStatusCode
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono
import java.time.Duration

/**
 * Prepare Qdrant collection if needed.
 */
@Component
class QdrantCollectionInitializer(
    private val properties: RagDialogueProperties,
    webClientBuilder: WebClient.Builder,
) : ApplicationRunner {
    private val log = LoggerFactory.getLogger(QdrantCollectionInitializer::class.java)
    private val webClient: WebClient

    init {
        val qdrant = properties.qdrant
        var builder = webClientBuilder.baseUrl(trimTrailingSlash(qdrant.url))
        if (StringUtils.hasText(qdrant.apiKey)) {
            builder = builder.defaultHeader("api-key", qdrant.apiKey)
        }
        webClient = builder.build()
    }

    override fun run(args: ApplicationArguments) {
        val qdrant = properties.qdrant
        if (!qdrant.autoCreateCollection) {
            log.info(
                "Qdrant auto-create is disabled. Skip collection init. collection={}",
                qdrant.collectionName,
            )
            return
        }

        if (collectionExists(qdrant.collectionName)) {
            log.info("Qdrant collection already exists. collection={}", qdrant.collectionName)
            return
        }

        createCollection(qdrant.collectionName, qdrant.vectorDimension)
    }

    private fun collectionExists(collectionName: String): Boolean {
        val exists =
            webClient
                .get()
                .uri("/collections/{collectionName}", collectionName)
                .exchangeToMono { response ->
                    val status: HttpStatusCode = response.statusCode()
                    when {
                        status.is2xxSuccessful -> {
                            Mono.just(true)
                        }

                        status.value() == 404 -> {
                            Mono.just(false)
                        }

                        else -> {
                            response
                                .bodyToMono(String::class.java)
                                .defaultIfEmpty("")
                                .flatMap { body ->
                                    Mono.error<Boolean>(
                                        IllegalStateException(
                                            "Qdrant collection lookup failed status=${status.value()} body=$body"
                                        )
                                    )
                                }
                        }
                    }
                }

        return exists.blockOptional(REQUEST_TIMEOUT).orElse(false)
    }

    private fun createCollection(
        collectionName: String,
        vectorDimension: Int,
    ) {
        val payload: Map<String, Map<String, Any>> =
            mapOf(
                "vectors" to
                    mapOf(
                        "size" to vectorDimension,
                        "distance" to "Cosine",
                    ),
            )

        webClient
            .put()
            .uri("/collections/{collectionName}", collectionName)
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(payload)
            .exchangeToMono { response ->
                val status = response.statusCode()
                when {
                    status.is2xxSuccessful || status.value() == 409 -> {
                        Mono.empty<Void>()
                    }

                    else -> {
                        response
                            .bodyToMono(String::class.java)
                            .defaultIfEmpty("")
                            .flatMap { body ->
                                Mono.error<Void>(
                                    IllegalStateException(
                                        "Qdrant collection create failed status=${status.value()} body=$body"
                                    )
                                )
                            }
                    }
                }
            }.block(REQUEST_TIMEOUT)

        log.info(
            "Qdrant collection is ready. collection={}, vectorDimension={}",
            collectionName,
            vectorDimension,
        )
    }

    private fun trimTrailingSlash(url: String): String {
        if (!StringUtils.hasText(url)) {
            return ""
        }
        if (url.endsWith("/")) {
            return url.substring(0, url.length - 1)
        }
        return url
    }

    companion object {
        private val REQUEST_TIMEOUT: Duration = Duration.ofSeconds(10)
    }
}
